File: frontend/src/services/query.ts
/**
 * Query execution API.
 */

import apiClient, { ensureCsrfToken } from './api'

export interface QueryTemplateParamSchema {
  type: 'integer' | 'string' | 'number'
  default?: unknown
  description?: string
}

export interface QueryTemplate {
  id: string
  name: string
  description: string
  cypher: string
  params: Record<string, unknown>
  param_schema: Record<string, QueryTemplateParamSchema>
}

export interface QueryExecuteRequest {
  graph: string
  cypher: string
  params?: Record<string, unknown>
  options?: Record<string, unknown>
  for_visualization?: boolean
  mutation_confirmed?: boolean
}

export interface QueryResultRow {
  data: Record<string, unknown>
}

export interface QueryExecuteResponse {
  columns: string[]
  rows: QueryResultRow[]
  row_count: number
  command?: string
  stats?: Record<string, unknown>
  request_id: string
  graph_elements?: {
    nodes: Array<{
      id: string
      label: string
      properties: Record<string, unknown>
      type: string
    }>
    edges: Array<{
      id: string
      label: string
      source: string
      target: string
      properties: Record<string, unknown>
      type: string
    }>
    paths?: Array<{
      type: string
      node_ids?: string[]
      edge_ids?: string[]
      segments?: unknown[]
    }>
  }
  visualization_warning?: string
}

export interface QueryCancelResponse {
  cancelled: boolean
  request_id: string
}

export interface QueryStreamRequest {
  graph: string
  cypher: string
  params?: Record<string, unknown>
  chunk_size?: number
  offset?: number
}

export interface QueryStreamChunk {
  columns: string[]
  rows: QueryResultRow[]
  chunk_size: number
  offset: number
  has_more: boolean
  total_rows?: number | null
}

export const queryAPI = {
  listTemplates: async (): Promise<QueryTemplate[]> => {
    const response = await apiClient.get<QueryTemplate[]>('/queries/templates')
    return response.data
  },

  execute: async (request: QueryExecuteRequest): Promise<QueryExecuteResponse> => {
    const response = await apiClient.post<QueryExecuteResponse>(
      '/queries/execute',
      request
    )
    return response.data
  },

  cancel: async (requestId: string, reason?: string): Promise<QueryCancelResponse> => {
    const response = await apiClient.post<QueryCancelResponse>(
      `/queries/${requestId}/cancel`,
      { reason }
    )
    return response.data
  },

  stream: async function* (
    request: QueryStreamRequest,
    signal?: AbortSignal
  ): AsyncGenerator<QueryStreamChunk, void, unknown> {
    const csrfToken = (await ensureCsrfToken()) || sessionStorage.getItem('csrf_token') || ''
    const response = await fetch('/api/v1/queries/stream', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-CSRF-Token': csrfToken,
      },
      credentials: 'include',
      body: JSON.stringify(request),
      signal,
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.error?.message || 'Streaming failed')
    }

    const reader = response.body?.getReader()
    if (!reader) {
      throw new Error('Response body is not readable')
    }

    const decoder = new TextDecoder()
    let buffer = ''

    try {
      while (true) {
        if (signal?.aborted) break
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (line.trim()) {
            try {
              const chunk = JSON.parse(line) as QueryStreamChunk
              yield chunk
            } catch (e) {
              console.error('Failed to parse chunk:', e, line)
            }
          }
        }
      }

      // Process remaining buffer
      if (buffer.trim() && !signal?.aborted) {
        try {
          const chunk = JSON.parse(buffer) as QueryStreamChunk
          yield chunk
        } catch (e) {
          console.error('Failed to parse final chunk:', e)
        }
      }
    } finally {
      reader.releaseLock()
    }
  },
}

File: backend/app/models/query.py
"""Query execution models."""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class QueryExecuteRequest(BaseModel):
    """Request to execute a Cypher query."""

    graph: str = Field(..., description="Graph name")
    cypher: str = Field(..., description="Cypher query string")
    params: Optional[Dict[str, Any]] = Field(
        default=None, description="Query parameters"
    )
    options: Optional[Dict[str, Any]] = Field(
        default=None, description="Execution options"
    )
    for_visualization: bool = Field(
        default=False,
        description="If true, LIMIT is added to Cypher when absent to cap result size",
    )


class QueryResultRow(BaseModel):
    """A single row in query result."""

    data: Dict[str, Any] = Field(..., description="Row data")


class QueryExecuteResponse(BaseModel):
    """Response from query execution."""

    columns: List[str] = Field(..., description="Column names")
    rows: List[QueryResultRow] = Field(..., description="Result rows")
    row_count: int = Field(..., description="Number of rows")
    command: Optional[str] = Field(
        default=None, description="PostgreSQL command type"
    )
    stats: Optional[Dict[str, Any]] = Field(
        default=None, description="Query execution statistics"
    )
    request_id: str = Field(..., description="Request ID for cancellation")
    graph_elements: Optional[Dict[str, List[Dict[str, Any]]]] = Field(
        default=None,
        description="Extracted graph elements (nodes, edges) for visualization",
    )
    visualization_warning: Optional[str] = Field(
        default=None,
        description="Warning message if result exceeds visualization limits",
    )


class QueryCancelRequest(BaseModel):
    """Request to cancel a query."""

    reason: Optional[str] = Field(default=None, description="Cancellation reason")


class QueryCancelResponse(BaseModel):
    """Response from query cancellation."""

    cancelled: bool
    request_id: str


class QueryStreamRequest(BaseModel):
    """Request to stream query results."""

    graph: str = Field(..., description="Graph name")
    cypher: str = Field(..., description="Cypher query string")
    params: Optional[Dict[str, Any]] = Field(
        default=None, description="Query parameters"
    )
    chunk_size: int = Field(
        default=1000, ge=1, le=10000, description="Number of rows per chunk"
    )
    offset: int = Field(
        default=0, ge=0, description="Offset for pagination"
    )


class QueryStreamChunk(BaseModel):
    """A chunk of query results."""

    columns: List[str] = Field(..., description="Column names")
    rows: List[QueryResultRow] = Field(..., description="Result rows in this chunk")
    chunk_size: int = Field(..., description="Number of rows in this chunk")
    offset: int = Field(..., description="Offset of this chunk")
    has_more: bool = Field(..., description="Whether more rows are available")
    total_rows: Optional[int] = Field(
        default=None, description="Total number of rows (if known)"
    )

